src/Parser/Core/Modules/CastEfficiency.js

```diff
@@ -4,182 +4,210 @@ import SpellLink from 'common/SpellLink';
-      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
-      getCooldown: haste => 60,
-      isActive: combatant => combatant.hasTrinket(ITEMS.TOME_OF_UNRAVELING_SANITY.id),
-    },
-    {
-      spell: SPELLS.VELENS_FUTURE_SIGHT_BUFF,
-      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
-      getCooldown: haste => 75,
-      isActive: combatant => combatant.hasTrinket(ITEMS.VELENS_FUTURE_SIGHT.id),
-    },
-    {
-      spell: SPELLS.SUMMON_DREAD_REFLECTION,
-      name: `${ITEMS.SPECTER_OF_BETRAYAL.name}`,
-      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
-      getCooldown: haste => 45,
-      isActive: combatant => combatant.hasTrinket(ITEMS.SPECTER_OF_BETRAYAL.id),
-    },
-    {
-      spell: SPELLS.CEASELESS_TOXIN,
-      name: `${ITEMS.VIAL_OF_CEASELESS_TOXINS.name}`,
-      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
-      getCooldown: haste => 60, // TODO: add detection if target has died and reduced cooldown
-      isActive: combatant => combatant.hasTrinket(ITEMS.VIAL_OF_CEASELESS_TOXINS.id),
-    },
-  ];
-
-  getAbility(spellId) {
-    return this.constructor.CPM_ABILITIES.find(ability => {
-      if (ability.spell.id === spellId) {
-        return !ability.isActive || ability.isActive(this.combatants.selected);
-      }
-      return false;
-    });
+
+  /*
+   * Gets the number of ms the given spell has been on CD since the beginning of the fight.
+   * Only works on spells entered into CastEfficiency list.
+   */
+  _getTimeOnCooldown(spellId) {
+    const history = this.spellHistory.historyBySpellId[spellId];
+    if(!history) {
+      return;
+    }
+
+    let lastBeginTimestamp = null;
+    const nonEndTimeOnCd = history
+        .filter(event => event.type === 'updatespellusable')
+        .reduce((acc, event) => {
+          if(event.trigger === 'begincooldown') {
+            lastBeginTimestamp = event.timestamp;
+            return acc;
+          } else if(event.trigger === 'endcooldown') {
+            const timeOnCd = (event.timestamp - lastBeginTimestamp) || 0;
+            lastBeginTimestamp = null;
+            return acc + timeOnCd;
+          } else {
+            return acc;
+          }
+        }, 0);
+    const endTimeOnCd = (!lastBeginTimestamp) ? 0 : this.owner.currentTimestamp - lastBeginTimestamp;
+    return nonEndTimeOnCd + endTimeOnCd;
   }
-  getExpectedCooldownDuration(spellId) {
-    const ability = this.getAbility(spellId);
-    return ability ? (ability.getCooldown(this.haste.current, this.combatants.selected) * 1000) : undefined;
+
+  /*
+   * Gets the average number of ms the given spell took to cooldown (or generate a charge)
+   * Only works on spells entered into CastEfficiency list.
+   */
+  _getAverageCooldown(spellId) {
+    const history = this.spellHistory.historyBySpellId[spellId];
+    if(!history) {
+      return;
+    }
+
+    let lastStartChargeTimestamp = null;
+    let recharges = 0;
+    const totalRechargingTime = history
```

Comment:
What's the difference between this and `_getTimeOnCooldown`?

In any case it should probably be its own method.

Reply:
TimeOnCooldown gets total time on cooldown, up to and including spell still cooling down at the current time.

AverageCooldown gets the average time it takes spell to recharge, disregards a cooldown that's in progress at the current time.

I considered trying to merge more of the functionality, but it seemed like I was twisting things too much so I made them seperate methods.

Reply:
I meant the statement this comment was one; it gets the totalRechargingTime which should be equal to _getTimeOnCooldown so could it be removed?

Reply:
Oh, this totalRechargingTime includes hooks to bump recharge count, and also disregards the time of the current cooldown. There is probably room to merge some code.